python/whylogs/extras/nlp_metric.py

```diff
@@ -0,0 +1,289 @@
+from dataclasses import dataclass
+from io import BytesIO
+from typing import Any, Dict, List, Tuple
+
+import numpy as np
+
+from whylogs.core.configs import SummaryConfig
+from whylogs.core.metrics.compound_metric import CompoundMetric
+from whylogs.core.metrics.deserializers import deserializer
+from whylogs.core.metrics.metric_components import (
+    FractionalComponent,
+    IntegralComponent,
+    MetricComponent,
+)
+from whylogs.core.metrics.metrics import (
+    DistributionMetric,
+    FrequentItemsMetric,
+    Metric,
+    MetricConfig,
+    OperationResult,
+)
+from whylogs.core.metrics.serializers import serializer
+from whylogs.core.preprocessing import ListView, PreprocessedColumn
+from whylogs.core.proto import MetricComponentMessage, MetricMessage
+
+_SMALL = np.finfo(float).eps
+
+
+def _reciprocal(s: np.ndarray) -> np.ndarray:
+    """Return pseudoinverse of singular value vector"""
+    # should also zap if too small relative to s[0]
+    return np.array([1 / x if x > _SMALL else 0 for x in s])
+
+
+def _serialize_ndarray(a: np.ndarray) -> bytes:
+    bio = BytesIO()
+    np.save(bio, a, allow_pickle=False)
+    return bio.getvalue()
+
+
+def _deserialize_ndarray(a: bytes) -> np.ndarray:
+    bio = BytesIO(a)
+    return np.load(bio, allow_pickle=False)
+
+
+class VectorComponent(MetricComponent[np.ndarray]):
+    # mtype = np.ndarray
+    type_id = 101
+
+
+@serializer(type_id=101)
+def serialize(value: np.ndarray) -> MetricComponentMessage:
+    return MetricComponentMessage(serialized_bytes=_serialize_ndarray(value))
+
+
+@deserializer(type_id=101)
+def deserialize(msg: MetricComponentMessage) -> "VectorComponent":
+    return VectorComponent(value=_deserialize_ndarray(msg.serialized_bytes))
+
+
+"""
+    def to_protobuf(self) -> MetricComponentMessage:
+        return MetricComponentMessage(type_id=self.type_id, serialized_bytes=_serialize_ndarray(self.value))
+
+    @classmethod
+    def from_protobuf(cls, msg: MetricComponentMessage, registries: Optional[Registries] = None) -> "VectorComponent":
+        return VectorComponent(value=_deserialize_ndarray(msg.serialized_bytes))
+"""
+
+
+@dataclass(frozen=True)
+class SvdMetricConfig(MetricConfig):
+    k: int = 0
+    decay: float = 1.0
+
+
+# The SvdMetric classes just hold (and optionally update) the SVD
+# sketch. They can compute residuals from the current SVD approximation,
+# but they do not maintain any statistics about the distribution. See
+# the NlpMetric for tracking the residual distribution.
+
+
+@dataclass(frozen=True)
+class SvdMetric(Metric):
```

Comment:
I like the metric factoring with separating the svd from the nlp metric, can we add some tests for this and NLPMetric?